README.md
![alt text](https://github.com/Steveb1968/script.pseudotv/blob/master/resources/images/Default.png?raw=true "PseudoTV Logo")

PseudoTV for Kodi
======

[View branch download information.](#branches-guide)


![screenshot](https://github.com/Steveb1968/script.pseudotv/blob/master/resources/screenshots/screenshot-01.png?raw=true)

### What is it?
It's channel-surfing for your media center. Never again will you have to actually pick what you want to watch. Use an electronic program guide (EPG) to view what's on or select a show to watch. This script will let you create your own channels and, you know, watch them. Doesn't actually sound useful when I have to write it in a readme. Eh, try it and decide if it works for you.

---
### Requirements
This **ONLY** uses your existing video library to play video. It will **NOT** play video from the internet. If you do not have a significant number of videos in your library, then this script probably isn't for you. Sorry.

---
### Features

* Automatic channel creation based on your library.
* Optionally customize the channels you want with the channel configuration tool.
* Utilize the Kodi smart playlist editor to create advanced channel setups.
* Use an EPG and view what was on, is on, and will be on. Would you rather see something that will be on later? Select it and watch it now!
* Want to pause a channel while you watch another? And then come back to it and have it be paused still? Sounds like a weird request to me, but if you want to do it you certainly can. It's a feature!
* An idle-timer makes sure you aren't spinning a hard-drive needlessly all night.
* Discover the other features on your own (so that I don't have to list them all...I'm lazy).

---
### Setup
1. First, install it.  This is self-explanatory (hopefully).  Really, that's all that is necessary.  Default channels will be created without any intervention.  You can choose to setup channels (next step) if you wish.
2. _Instructions to create your own channels:_ Inside of the addon config, you may open the channel configuration tool. Inside of here you can select a channel to modify. You may then select it's type and any options. For a basic setup, that's all you need to do. It's worth noting that you may create a playlist using the smart playlist editor and then select that playlist in the channel config tool (Custom Playlist channel type). Additionally, you may select to add advanced rules to certain channels. There are quite a few rules that are currently available, and hopefully they should be relatively self-explanitory.

	This is a readme and should include descriptions of them all... who knows, maybe it will some day.

---
### Controls
There are only a few things you need to know in order to control everything. First of all, the Stop button ('X') stops the video and exits the script. You may also press the Previous Menu ('Escape\Back') button to do this (don't worry, it will prompt you to verify first). Scroll through channels using the arrow up and down keys, or alternatively by pressing Page up or down. To open the EPG, press the Select key ('Enter'). Move around using the arrow keys. Start a program by pressing Select. Pressing Previous Menu ('Escape\Back') will close the EPG. Press the info key ('I') to display or hide the info window.  When it is displayed, you can look at the previous and next shows on this channel using the arrow keys left and right. To access the video osd window press the context menu key ('C\Menu'), to exit the osd press ('Escape\Back').

---
### Settings

**General Settings -**

* **Configure Channels:** This is the channel configuration tool.  From here you can modify the settings for each individual channel.    
* **Force Channel Reset:** If you want your channels to be reanalyzed then you can turn this on.
* **Auto-off Timer:** The amount of time (in minutes) of idle time before the script is automatically stopped.
* **Time Between Channel Resets:** This is how often your channels will be reset. Generally, this is done automatically based on the duration of the individual channels and how long they've been watched. You can change this to reset every certain time period (day, week, month).
* **Default channel times at startup:** This affects where the channels start playing when the script starts.  Resume will pick everything up where it left off. Random will start each channel in a random spot. Real-Time will act like the script was never shut down, and will play things at the time the EPG said they would play.
* **Background Updating:** The script uses multiple threads to keep channels up to date while other channels are playing. In general, this is fine. If your computer is too slow, though, it may cause stuttering in the video playback. This setting allows you to minimize or disable the use of these background threads.
* **Enable Channel Sharing:** Share the same configuration and channel list between multiple computers. If you're using real-time mode (the default) then you can stop watching one show on one computer and pick it up on the other. Or you can have both computers playing the same lists at the same time.
* **Shared Channels Folder:** If channel sharing is enabled, this is the location available from both computers that will keep the settings and channel information.


**Visual Settings -**

* **Info when Changing Channels:** Shows the current media information when changing channels. The duration of the info can be set in the sub-setting "Changing channel info duration".
* **Always show channel watermark:** Always display the current channel logo watermark. The position of the logo can be adjusted via the sub-setting "Set watermark position" (upper left, upper right, lower right, lower left).
* **Hide year and episode information:** Removes the year (movies) and SxEP (episodes). A force channel reset is needed for the setting to take effect.  
* **Always show channel logo in epg grid:** Shows the channel logo's in the epg grid.
* **Channel Logo Folder:** The place where channel logos are stored.
* **Clock Display:** Select between a 12-hour or 24-hour clock in the EPG.
* **Show Coming Up Next box:** A little box will notify you of what's coming up next when the current show is nearly finished.
* **Hide very short videos:** Don't show clips shorter than the "Duration of Short videos" setting. Effects the EPG, coming up next box and the info box. This is helpful if you use bumpers or commercials.


**Tweaks -**

* **Playlist Media Limit:** Limit the playlist items in a channel generated by pseudotv. Smaller values will result in quicker load/rebuild times.
* **OSD Channel Number Color:** Change the color of the channel number located at the top left, seen when changing channels and on startup.
* **Seek step forward:** Option to adjust the seek step forward (right arrow key in fullscreen video). Options include 10 sec,30 sec,60 sec,3 min,5 min,10 min,30 min.
* **Seek step backward:** Option to adjust the seek step backward (left arrow key in fullscreen video). Options include -10 sec,-30 sec,-60 sec,-3 min,-5 min,-10 min,-30 min.
* **Reset Watched Status:** Option to reset watched status and resume points of videos while watching PseudoTV.


**Auto Start -**

* **Activate Service:** Activate auto start. Pseudotv will automatically start when kodi is started.
* **Service Delay:** Delay the auto start of Pseudotv at Kodi start-up. This is useful for low end hardware or if your skin loads multi-pal scripts at start-up.
* **Show Notification:** Self explanatory I hope. Will show a notification of Pseudotv's attempt to auto-start.

---
### Addon boolean condition

**&lsaquo;visible&rsaquo;String.IsEmpty(Window(home).Property(PseudoTVRunning))&lsaquo;/visible&rsaquo;**  
Useful for hiding skin xml files such as DialogBusy.xml/DialogSeekBar.xml 
 
**Tip: DialogSeekBar.xml**  
Replace *"Player.DisplayAfterSeek"* with  
*"[Player.DisplayAfterSeek\+String.IsEmpty(Window(home).Property(PseudoTVRunning))]"*    

---
### Branches guide

* **Master branch:** Suitable for Kodi VER:18 and above.
* **Leia branch:** Suitable for Kodi VER:18.
* **Krypton branch:** Suitable for Kodi VER:17 and below. 

---
### Credits

**Developer:** Jason102, Steveb.<br>
**Code Additions:** Sranshaft, TheOddLinguist, Canuma, rafaelvieiras, fnord12.<br>
**Skins:** Sranshaft, Zepfan, Steveb.<br>
**Preset Images:** Jtucker1972.<br>
**Languages:** CyberXaz, Machine-Sanctum, rafaelvieiras, Eng2Heb.

changelog.txt
2.4.5 (steveb)
    Hebrew translation. Thanks 'Eng2Heb'
    Added year for movie info.

2.4.4 (steveb)
    Added option to enable/disable channel logo's in the epg grid.
    New Default Skin (Estuary based).
    Added default font.

2.4.3 (steveb)
    Check for playing media.
    Improve timebar position.
    Code cleanup.
    Cleanup DialogProgress

2.4.2 (steveb)
    Jarvis code cleanup

2.4.1 (steveb)
    Added option to set seek times forward/backward (Isengard & above).
    Code cleanup.

2.4.0 (steveb)
    Helix support.

2.3.0 (steveb)
    Automatic conversion of colour channel logo's to grey-scale for channel bug.

2.2.0 (steveb)
    Gotham Support.
    Code Cleanup.
    Localize Strings.
    Changed max mu3 string length from 4096(512 bytes) to 16384(2048 bytes).

2.1.0
    New rule: reset every x days
    Update to Night skin (thanks Steveb).
    Better exception handling.
    Change 'short clip' length.
    Include IceLibrary streams.
    Configuration bug fixes.
    Addition of Metropolis skin (thanks Steveb).
    Updates to skins (thanks again Steveb).
    Fix for color truncation (thanks Josebiro).
    Fix for missing texture (thanks lambda81).
    Hide channel bug when changing channel.
    Properly detect already running.
    Option to include the season / episode in the show title.

2.0.0
    New advanced channel rule system.
    Share channels between multiple systems.
    Addition of Dutch language (thanks Machine-Sanctum).
    Support for MOV file parsing.
    Multiple thread deletion issues fixed.
    Background channel loading.
    EPG bugs fixed.
    Support for directory-type channels.
    Select mode for background channel thread.
    EPG drawing optimization.
    Select between 12-hour and 24-hour clock.
    Better channel creation for first-time users.
    Much better time accuracy between restarts.
    Added Xeebo skin (thanks Steveb).

1.2.1
    Now ask the user to enable the webserver if it isn't on.
    Fix to use the proper webserver port.
    Update to the Night skin (thanks Steveb).
    Added the Simplicity skin (thanks Zepfan).
    Several small performance improvements.
    Shift 30 seconds forward or backward by using the left and right arrow keys.
    Many random fixes.

1.2.0
    The user can now specify how channels start: resume, follow the EPG (real-
        time), or random.
    Background updating of channels.
    Fixed issues when used in Eden.
    No longer sometimes crashes on exit.
    Turns off repeat when exiting.
    Added Spanish translation (thanks CyberXaz).
    Updated the Night skin (thanks Steveb).

1.1.0
    Added the channel configuration utility.
    Fixed a potential crash in the mp4 parser.
    Separated the channel creation code into a different module.
    Changed a few of the configuration strings.
    Added the ability to refresh channels on every start.
    Improved the channel changing speed.
    Resume a channel at the proper time when switching between channels.
    Allow the user to specify a directory for channel logos.
    Allow the use of the OSD.
    Added an FLV parser.
    New skins from Zepfan.
    Fixed a problem with shows not being resumed properly from the EPG.
    Gave skinners the ability to change the EPG button text color.
    Updates to nearly all skins (thanks Zepfan and Sranshaft).

1.0.1
    Able to use the Previous Menu key to exit.
    Update to the Refocus skin from Sranshaft.

1.0.0
    Crazy jump in version to 1.0 because I can.
    Turn off shuffling if it's on.
    Playlist parsing is completely minidom now.
    Manually shuffle playlists when the order is random.
    Allow the stop button inside of the 2 second window.
    Added the Confluence skin (thanks Sranshaft).
    Added the Transparency skin (thanks Zepfan).
    Now have channel icons for most presets (thanks jtucker1972).
    Fixed an uncaught exception in video parsers.
    Works if onClick isn't called in the EPG.

0.2.0
    New default skin (thanks to Sranshaft)
    Added an info window (again, Sranshaft).
    There is now a PseudoTV icon (thanks bridgman).
    Channel icons.
    Control over button placement put into the skin.
    Start the script on the previously viewed channel.
    EPG speed improvement.
    Support mixed playlists.
    Support for future JSON modification.

0.1.5
    Resolved the problem with not resuming the channel properly.
    Fixed the issue introduced in 0.1.4 that prevending gathering information
        on Windows systems.
    Put several of the controls into the hands of the skinners.

0.1.4
    New method of scanning channels (thanks to TheOddLinguist) that correctly
        uses movie sets and TV show-type playlists.
    Display music video information in the EPG.

0.1.3
    There was an issue where XBMC woukd sometimes return a float when an int
        was expected, resulting in a type error and a freeze.
    Sometimes the EPG wouldn't properly select or display the correct show.
    The EPG would sometimes stop scrolling.
    Changed the EPG to left-shift the time offsets.

0.1.2
    Windows systems experienced freezes...but no more!
    There was also an issue getting the duration of an episode.

0.1.1
    Fixed a freeze when there are unexpected things in the smart playlist.   

0.1.0 (Initial alpha version)


resources/lib/Globals.py
#   Copyright (C) 2011 Jason Anderson
#
#
# This file is part of PseudoTV.
#
# PseudoTV is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# PseudoTV is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with PseudoTV.  If not, see <http://www.gnu.org/licenses/>.

import os
import xbmcaddon, xbmc, xbmcgui, xbmcvfs
import Settings

from FileAccess import FileLock

def debug(msg, *args):
    try:
        txt=u''
        msg=unicode(msg)
        for arg in args:
            if type(arg) == int:
                arg = unicode(arg)
            if type(arg) == list:
                arg = unicode(arg)
            txt = txt + u"/" + arg
        if txt == u'':
            xbmc.log(u"PSTV: {0}".format(msg).encode('ascii','xmlcharrefreplace'), xbmc.LOGDEBUG)
        else:
            xbmc.log(u"PSTV: {0}#{1}#".format(msg, txt).encode('ascii','xmlcharrefreplace'), xbmc.LOGDEBUG)
    except:
        print "PSTV: Error in Debugoutput"
        print msg
        print args

ADDON = xbmcaddon.Addon(id='script.pseudotv')
ADDON_ID = ADDON.getAddonInfo('id')
ADDON_NAME = ADDON.getAddonInfo('name')
LANGUAGE = ADDON.getLocalizedString
CWD = ADDON.getAddonInfo('path').decode("utf-8")
VERSION = ADDON.getAddonInfo('version')
ICON = ADDON.getAddonInfo('icon')

def log(msg, level = xbmc.LOGDEBUG):
    try:
        xbmc.log(ADDON_ID + '-' + ascii(msg), level)
    except:
        pass


def uni(string, encoding = 'utf-8'):
    if isinstance(string, basestring):
        if not isinstance(string, unicode):
           string = unicode(string, encoding)

    return string

def ascii(string):
    if isinstance(string, basestring):
        if isinstance(string, unicode):
           string = string.encode('ascii', 'ignore')

    return string


TIMEOUT = 15 * 1000
PREP_CHANNEL_TIME = 60 * 60 * 24 * 5
NOTIFICATION_CHECK_TIME = 5
NOTIFICATION_TIME_BEFORE_END = 90
NOTIFICATION_DISPLAY_TIME = 8

MODE_RESUME = 1
MODE_ALWAYSPAUSE = 2
MODE_ORDERAIRDATE = 4
MODE_RANDOM = 8
MODE_REALTIME = 16
MODE_STARTMODES = MODE_RANDOM | MODE_REALTIME | MODE_RESUME

SETTINGS_LOC = ADDON.getAddonInfo('profile').decode("utf-8")
CHANNEL_SHARING = False
LOCK_LOC = xbmc.translatePath(os.path.join(SETTINGS_LOC, 'cache' + '/'))

if ADDON.getSetting('ChannelSharing') == "true":
    CHANNEL_SHARING = True
    LOCK_LOC = xbmc.translatePath(os.path.join(ADDON.getSetting('SettingsFolder'), 'cache' + '/'))

IMAGES_LOC = xbmc.translatePath(os.path.join(CWD, 'resources', 'images' + '/'))
LOGOS_LOC = xbmc.translatePath(os.path.join(CWD, 'resources', 'logos' + '/'))
CHANNELS_LOC = os.path.join(SETTINGS_LOC, 'cache' + '/')
GEN_CHAN_LOC = os.path.join(CHANNELS_LOC, 'generated' + '/')
MADE_CHAN_LOC = os.path.join(CHANNELS_LOC, 'stored' + '/')
CHANNELBUG_LOC = xbmc.translatePath(os.path.join(CHANNELS_LOC, 'ChannelBug' + '/'))

CHANNELBUG_POS =[[19,19],[1695,19],[1695,952],[19,952]]

SHORT_CLIP_ENUM = [15, 30, 60, 90, 120, 180, 240, 300, 360]
INFO_DUR = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
SEEK_FORWARD = [10, 30, 60, 180, 300, 600, 1800]
SEEK_BACKWARD = [-10, -30, -60, -180, -300, -600, -1800]
MEDIA_LIMIT = [10, 25, 50, 100, 250, 500, 1000, 0]
CHANNEL_DELAY = [25, 50, 100, 200, 300, 400, 500, 600, 700, 800, 900, 1000]
HOLD_ACTIONS = [.25, .5, 1, 1.5, 2, 2.5, 3, 3.5, 4]
ASSIGNED_DURATION = [1, 5, 10, 15, 30, 45, 60, 90, 120]
DIR_ASSIGNED_DURATION = [1, 5, 10, 15, 30, 45, 60, 90, 120]
BUG_BRIGHTNESS = [.25, .5, .75, 1.0, 1.25, 1.5, 1.75, 2.0, 2.25, 2.5, 2.75, 3.0]


NUM_COLOUR = ['0xFFFF0000', '0xFF00FF00', '0xFF0000FF', '0xFFFFFF00', '0xFF00FFFF', '0xFFFFA500', '0xFFFF00FF', '0xFF808080', '0xFFFFFFFF']

GlobalFileLock = FileLock()
ADDON_SETTINGS = Settings.Settings()

TIME_BAR = 'pstvTimeBar.png'
BUTTON_NO_FOCUS = 'pstvButtonNoFocus.png'
BUTTON_FOCUS = 'pstvButtonFocus.png'

RULES_ACTION_START = 1
RULES_ACTION_JSON = 2
RULES_ACTION_LIST = 4
RULES_ACTION_BEFORE_CLEAR = 8
RULES_ACTION_BEFORE_TIME = 16
RULES_ACTION_FINAL_MADE = 32
RULES_ACTION_FINAL_LOADED = 64
RULES_ACTION_OVERLAY_SET_CHANNEL = 128
RULES_ACTION_OVERLAY_SET_CHANNEL_END = 256

# Maximum is 10 for this
RULES_PER_PAGE = 7

ACTION_MOVE_LEFT = 1
ACTION_MOVE_RIGHT = 2
ACTION_MOVE_UP = 3
ACTION_MOVE_DOWN = 4
ACTION_PAGEUP = 5
ACTION_PAGEDOWN = 6
ACTION_SELECT_ITEM = 7
ACTION_SELECT_ITEM2 = 100 #Mouse Left Click
ACTION_PREVIOUS_MENU = (9, 10, 92, 216, 247, 257, 275, 61467, 61448,)
ACTION_SHOW_INFO = 11
ACTION_STOP = 13
ACTION_OSD = 117
ACTION_NUMBER_0 = 58
ACTION_NUMBER_9 = 67
ACTION_INVALID = 999
ACTION_MOUSE_RIGHT_CLICK = 101 #Mouse Right Click
CONTEXT_MENU = 117
ACTION_PREV_PICTURE = 29
